src/plugin/svipc/sysvipcwrappers.cpp

```diff
@@ -183,9 +188,72 @@ shmdt(const void *shmaddr)
 // in order to install its own hooks. For us, shmdt() is the only interesting
 // one. Instead of giving the address of our wrapper to the hook library, we
 // want to return the address in libc. See PR #472 for details.
+static int dlsym_addr_instance = 0;
+static const char *dlsym_symbol = NULL;
+static void *dlsym_retval;
+// extern "C"
+static int
+callback(struct dl_phdr_info *info, size_t size, void *data) {
+  void *handle = dlopen(info->dlpi_name, RTLD_LAZY);
+  dlerror(); // initialize for errors
+  void *address = dlsym(handle, dlsym_symbol);
+  if (address == NULL && dlerror() != NULL) {
+    dlsym_retval = dlsym(handle, dlsym_symbol); // Set dlerror to error string
+    dlclose(handle);
+    return 0;
+  } else {
+    dlclose(handle);
+  }
+  if (address != NULL) {
+    dlsym_addr_instance++;
+    printf("Symbol %s in %s: found at address %p\n",
+           dlsym_symbol, info->dlpi_name, address);
+  }
+  if (dlsym_addr_instance == 1) {
+    printf("Caller of %s() in %s: wrapper fnc. address:  %p\n",
+           dlsym_symbol, info->dlpi_name, address);
+  }
+  if (dlsym_addr_instance == 2) {
+    printf("RTLD_NEXT:  address:  %p\n", address);
+    dlsym_retval = address;
+  }
+  return 0;
+}
+
+// extern "C"
+static void *
+dlsym_with_rtld_next(const char *symbol) {
+  dlsym_symbol = symbol;
+  dlsym_addr_instance = 0;
+  dlsym_retval = NULL;
+  dl_iterate_phdr(callback, NULL);
+  return dlsym_retval;
+}
+
+//   We need to wrap dlsym to disable and re-enable ckpt.  But then,
+// dlsym(RTLD_NEXT, ...) no longer works, since glibc:dlsym() looks up
+// the stack to find the caller's library, and discovers a DMTCP library
+// instead of the original target library.
+//   It was originally placed in the svipc plugin so that RTLD_NEXT would
+// point to glibc.  But some targets want RTLD_NEXT to point to a library
+// before glibc.  This happened when ompt_start_tool called dlsym()
+// to see if there was a preferred definition later in the search path.
+// By default, dlsym() should return NULL.  But now, dlsym is called by
+// DMTCP, and so dlsysm discovers the Intel OpenMP library (libiomp5.so)
+// later in the search path.
+// FIXME:  This code belongs in 'src' in libdmtcp.so, and not here.
```

Comment:
I agree. This code belongs in libdmtcp.so along with the code for dlopen (which currently sits in plugin/dl directory).

Reply:
I've created `src/dlwrappers.cpp` now, and moved the 'dlsym' wrapper to there.